cpp/examples/arrow/rapidjson_row_converter.cc

```diff
@@ -0,0 +1,590 @@
+      std::shared_ptr<arrow::Schema> schema) override {
+    std::unique_ptr<arrow::RecordBatchBuilder> batch_builder;
+    ARROW_ASSIGN_OR_RAISE(batch_builder,
+                          arrow::RecordBatchBuilder::Make(schema, pool_, rows.size()));
+
+    JsonValueConverter converter(rows);
+    for (int i = 0; i < batch_builder->num_fields(); ++i) {
+      std::shared_ptr<arrow::Field> field = schema->field(i);
+      arrow::ArrayBuilder* builder = batch_builder->GetField(i);
+      ARROW_RETURN_NOT_OK(converter.Convert(*field.get(), builder));
+    }
+
+    std::shared_ptr<arrow::RecordBatch> batch;
+    ARROW_ASSIGN_OR_RAISE(batch, batch_builder->Flush());
+
+    DCHECK_OK(batch->ValidateFull());
+    return batch;
+  }
+
+ protected:
+  arrow::MemoryPool* pool_;
+};  // DocumentToArrowConverter
+
+int main(int argc, char** argv) {
+  // Get sizes
+  int32_t num_rows = argc > 1 ? std::atoi(argv[1]) : 100;
+  int32_t batch_size = argc > 2 ? std::atoi(argv[2]) : 100;
+
+  //(Doc section: Convert to Arrow)
+  // Write JSON records
+  std::vector<std::string> json_records = {
+      R"({"pk": 1, "date_created": "2020-10-01", "data": {"deleted": true, "metrics": [{"key": "x", "value": 1}]}})",
+      R"({"pk": 2, "date_created": "2020-10-03", "data": {"deleted": false, "metrics": []}})",
+      R"({"pk": 3, "date_created": "2020-10-05", "data": {"deleted": false, "metrics": [{"key": "x", "value": 33}, {"key": "x", "value": 42}]}})"};
+
+  std::vector<rapidjson::Document> records;
+  records.reserve(num_rows);
+  for (int32_t i = 0; i < num_rows; ++i) {
+    rapidjson::Document document;
+    document.Parse(json_records[i % json_records.size()].c_str());
+    records.push_back(std::move(document));
+  }
+  // for (const std::string& json : json_records) {
+  //   rapidjson::Document document;
+  //   document.Parse(json.c_str());
+  //   records.push_back(std::move(document));
+  // }
+  for (const rapidjson::Document& doc : records) {
+    rapidjson::StringBuffer sb;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+    // At batch_size >= 495, we segfault here:
+    doc.Accept(writer);
+    std::cout << sb.GetString() << std::endl;
+  }
+  auto tags_schema = arrow::list(arrow::struct_({
+      arrow::field("key", arrow::utf8()),
+      arrow::field("value", arrow::int64()),
+  }));
+  auto schema = arrow::schema(
+      {arrow::field("pk", arrow::int64()), arrow::field("date_created", arrow::utf8()),
+       arrow::field("data", arrow::struct_({arrow::field("deleted", arrow::boolean()),
+                                            arrow::field("metrics", tags_schema)}))});
+
+  // Create converter
+  DocumentToArrowConverter to_arrow_converter(arrow::default_memory_pool());
+
+  // Convert records into a table
+  arrow::Result<std::shared_ptr<arrow::Table>> table_result =
+      to_arrow_converter.ConvertToTable(records, schema);
+  std::shared_ptr<arrow::Table> table = std::move(table_result).ValueOrDie();
```

Comment:
in this case why not just `table = *to_arrow_converter.ConvertToTable();`?

Reply:
or, we could factor out a `Status Main()` that gets called from `int main`

Reply:
I like the idea of a `Status Main()`, since that enables us to show use of result/status macros in the example code blocks 👍